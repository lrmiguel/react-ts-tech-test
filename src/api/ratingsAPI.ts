type RatingsMetadataType = {
  dataSource: string;
  extractDate: string;
  itemCount: number;
  returncode: string;
  totalCount: number;
  totalPages: number;
  pageSize: number;
  pageNumber: number;
};

type LinkType = {
  rel: string;
  href: string;
};

export type EstablishmentsType = {
  establishments: {}[];
  meta: RatingsMetadataType;
  links: LinkType[];
};

export type AuthorityType = {
  Name: string;
  LocalAuthorityId: number;
}

export type AuthoritiesType = {
  authorities: AuthorityType[];
  meta?: RatingsMetadataType;
  links?: LinkType[];
};

export function getEstablishmentRatings(
  pageNum: number
): Promise<EstablishmentsType> {
  return fetch(
    `http://api.ratings.food.gov.uk/Establishments/basic/${pageNum}/10`,
    { headers: { "x-api-version": "2" } }
  ).then((res) => res.json());
}

export function getEstablishmentsByAuthority(
  authority: string,
  pageNum: number
): Promise<EstablishmentsType> {
  return fetch(
    `http://api.ratings.food.gov.uk/Establishments?localAuthorityId=${authority}&pageSize=10&pageNumber=${pageNum}`,
    { headers: { "x-api-version": "2" } }
  ).then((res) => res.json())
}

export function getAuthorities(
  pageNum: number
): Promise<AuthoritiesType> {
  return fetch(
    `http://api.ratings.food.gov.uk/Authorities/basic/${pageNum}/10`,
    { headers: { "x-api-version": "2" } }
  ).then((res) => res.json())
}

export function getAuthoritiesByName(
  name: string,
): Promise<AuthoritiesType> {
  // neither pageSize or pageNumber parameters are being applied when sent to the API
  return fetch(
    `http://api.ratings.food.gov.uk/Authorities?name=${name}`,
    { headers: { "x-api-version": "2" } }
  ).then((res) => res.json())
}