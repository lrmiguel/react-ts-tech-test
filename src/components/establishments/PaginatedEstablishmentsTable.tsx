import { useState, useEffect } from "react";
import { EstablishmentsTable } from "./EstablishmentsTable";
import { EstablishmentsTableNavigation } from "./EstablishmentsTableNavigation";
import { getEstablishmentRatings, getEstablishmentsByAuthority } from "../../api/ratingsAPI";
import { useQuery } from "react-query";

const tableStyle = {
  background: "#82C7AF",
  color: "white",
  padding: "10px 20px",
  width: "50rem",
  height: "29rem",
};

interface EstablishmentsTableProps {
  authority?: string;
}

export const PaginatedEstablishmentsTable: React.FC<EstablishmentsTableProps> = ({ authority }) => {
  const [error, setError] =
    useState<{ message: string; [key: string]: string }>();
  const [pageNum, setPageNum] = useState(1);

  // Fetches data from API based on 'establishments', authority, pageNum 
  const { data, isLoading, isFetching } = useQuery(
    ['establishments', authority, pageNum],
    () => fetchEstablishmentRatings(authority), // Pass authority as filter param
    {
      keepPreviousData: true,
      refetchOnWindowFocus: false,
      staleTime: 5 * 60 * 1000,
      cacheTime: 10 * 60 * 1000,
      onError: (error: unknown) => {
        setError({ message: JSON.stringify(error) });
        console.error("Error occurred when fetching establishments", error);
      }
    }
  );
  const establishments = data ?? [];

  const [pageCount] = useState(100);

  useEffect(() => {
    setPageNum(1);
  }, [authority]);

  async function handlePreviousPage() {
    pageNum > 1 && setPageNum(pageNum - 1);
  }

  async function handleNextPage() {
    pageNum < pageCount && setPageNum(pageNum + 1);
  }
  
  async function fetchEstablishmentRatings(authority?: string): Promise<{}[]> {
    const results = (authority) ? await getEstablishmentsByAuthority(authority, pageNum) : await getEstablishmentRatings(pageNum);
    return results.establishments;
  }

  if (error) {
    return <div>Error: {error.message}</div>;
  } else {
    return (
      <div style={tableStyle}>
        <h2>Food Hygiene Ratings</h2>
        <EstablishmentsTable establishments={establishments} loading={isLoading || isFetching} />
        <EstablishmentsTableNavigation
          loading={isLoading || isFetching}
          pageNum={pageNum}
          pageCount={pageCount}
          onPreviousPage={handlePreviousPage}
          onNextPage={handleNextPage}
        />
      </div>
    );
  }
};
