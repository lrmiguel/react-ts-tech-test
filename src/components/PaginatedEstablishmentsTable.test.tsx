import React from "react";
import { render, screen, fireEvent, act } from "@testing-library/react";
import { PaginatedEstablishmentsTable } from "./PaginatedEstablishmentsTable";
import * as ratingsAPI from "../api/ratingsAPI";

jest.mock("../api/ratingsAPI");

const mockData = [
    { BusinessName: "McDonald's", RatingValue: "4" },
    { BusinessName: "Burger King", RatingValue: "4" },
];

(ratingsAPI.getEstablishmentRatings as jest.Mock).mockResolvedValue({ establishments: mockData });

describe("PaginatedEstablishmentsTable", () => {
    beforeEach(() => {
        (ratingsAPI.getEstablishmentRatings as jest.Mock).mockImplementation(() => {
            return new Promise((resolve) =>
                setTimeout(() => resolve({ establishments: mockData }), 100)
            );
        });
    });

    /* Snapshot Tests */
    it("renders the table with updated font size", async () => {
        const { container } = render(<PaginatedEstablishmentsTable />);
        await screen.findByText("McDonald's");
        expect(container).toMatchSnapshot();
    });
    
    /* Unit and Integration Tests */
    it("renders loading state initially", async () => {
        render(<PaginatedEstablishmentsTable />);
        expect(screen.getByText("Loading...")).toBeInTheDocument();
        expect(await screen.findByText("McDonald's")).toBeInTheDocument();
    });

    it("shows loading when navigating to next page", async () => {
        render(<PaginatedEstablishmentsTable />);
        await screen.findByText("McDonald's");
        fireEvent.click(screen.getByText("+"));
        expect(await screen.findByText("Loading...")).toBeInTheDocument();
        
        await act(() => new Promise((resolve) => setTimeout(resolve, 100)));
        
        expect(screen.queryByText("Loading...")).not.toBeInTheDocument();
        expect(await screen.findByText("McDonald's")).toBeInTheDocument();
    });

    it("shows loading when navigating to previous page", async () => {
        render(<PaginatedEstablishmentsTable />);
        await screen.findByText("McDonald's");

        fireEvent.click(screen.getByText("+"));
        expect(await screen.findByText("Loading...")).toBeInTheDocument();
        await act(() => new Promise((resolve) => setTimeout(resolve, 100)));
        expect(screen.queryByText("Loading...")).not.toBeInTheDocument();
        expect(await screen.findByText("McDonald's")).toBeInTheDocument();

        // Simulate going back to the previous page

        fireEvent.click(screen.getByText("-"));
        expect(await screen.findByText("Loading...")).toBeInTheDocument();
        await act(() => new Promise((resolve) => setTimeout(resolve, 100)));
        expect(screen.queryByText("Loading...")).not.toBeInTheDocument();
        expect(await screen.findByText("McDonald's")).toBeInTheDocument();
    });
})