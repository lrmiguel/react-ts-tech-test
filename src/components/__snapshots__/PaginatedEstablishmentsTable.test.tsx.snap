// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginatedEstablishmentsTable renders the table with updated font size 1`] = `
<div>
  <div
    style="background: rgb(130, 199, 175); color: white; padding: 10px 20px; width: 50rem; height: 29rem;"
  >
    <h2>
      Food Hygiene Ratings
    </h2>
    <table>
      <tbody>
        <tr>
          <th
            style="font-size: 20px; text-align: left; padding-bottom: 10px; padding-right: 30px;"
          >
            Fav
          </th>
          <th
            style="font-size: 20px; text-align: left; padding-bottom: 10px; width: 37.5rem;"
          >
            Business Name
          </th>
          <th
            style="font-size: 20px; text-align: left; padding-bottom: 10px;"
          >
            Rating Value
          </th>
        </tr>
        <tr
          style="font-size: 20px; height: 1.8rem;"
        >
          <td
            style="width: auto;"
          >
            <input
              aria-label="Favourite McDonald's"
              type="checkbox"
            />
          </td>
          <td>
            McDonald's
          </td>
          <td>
            4
          </td>
        </tr>
        <tr
          style="font-size: 20px; height: 1.8rem;"
        >
          <td
            style="width: auto;"
          >
            <input
              aria-label="Favourite Burger King"
              type="checkbox"
            />
          </td>
          <td>
            Burger King
          </td>
          <td>
            4
          </td>
        </tr>
      </tbody>
    </table>
    <nav>
      <button
        disabled=""
        style="margin: 10px 5px;"
        type="button"
      >
        -
      </button>
      1
      <button
        style="margin: 10px 5px;"
        type="button"
      >
        +
      </button>
    </nav>
  </div>
</div>
`;
